### PR TITLE
Correctly estimate the required space for string in `StyledBuffer::prepend`

### LINKED ISSUE #65119: Compiling a file with a non-ASCII name with `--error-format=short` enabled makes error messages contain excessive spaces after "filename:line:column"
Consider the following rust source (in `test.rs`):

```rust
fn main() {
    println!("hello world
}
```

It is obvious that it has syntax error. It also seems obvious that the generated error shouldn't depend on name of file except noting the name of file. However, it doesn't happen in practice:

```shell
$ rustc test.rs --error-format=short
test.rs:2:14: error: unterminated double quote string
error: aborting due to previous error
$ mv test.rs 'тест.rs'
$ rustc 'тест.rs' --error-format=short
тест.rs:2:14:     error: unterminated double quote string
error: aborting due to previous error
```

Note extra spaces after `тест.rs:2:14:`.

Rust version (although it is not strictly relevant, see below):

```
$ rustc -vV
rustc 1.38.0 (625451e37 2019-09-23)
binary: rustc
commit-hash: 625451e376bb2e5283fc4741caa0a3e8a2ca4d54
commit-date: 2019-09-23
host: x86_64-unknown-linux-gnu
release: 1.38.0
LLVM version: 9.0
```

The core issue lies in the [implementation of `StyledBuffer::prepend` method](https://github.com/rust-lang/rust/blob/950fe6686da1fdbaf25341cc5fafe1bb9fb1b180/src/librustc_errors/styled_buffer.rs#L112-L123). The idea of the code is to make enough space in the beginning of line (filled with spaces) and then overwrite it character by character. However, in order to estimate the required space, [it uses `.len()`](https://github.com/rust-lang/rust/blob/950fe6686da1fdbaf25341cc5fafe1bb9fb1b180/src/librustc_errors/styled_buffer.rs#L114) which is wrong since `str` in Rust can contain non-ASCII which require more than one byte for storage and `StyledBuffer` keeps lines in `Vec<char>`s. It means that this isuue reveals itself only when a non-ASCII string is passed as an argument — and that's exactly what's happening in [`EmitterWriter::emit_message_default`](https://github.com/rust-lang/rust/blob/950fe6686da1fdbaf25341cc5fafe1bb9fb1b180/src/librustc_errors/styled_buffer.rs#L114) under very specific circumstances: _when_ processing a primary file, _when_ `self.short_message` is set (i. e. `--error-message=short`) and _when_ name of the primary file happens to have a non-ASCII name.

The errorneous code for `StyledBuffer::prepend` was introduced in 2e8e73cb951e2095e444c5cc91403f7e9c5f1065. This bug landed in stable [in 1.28.0](https://github.com/rust-lang/rust/blob/master/RELEASES.md#compiler-9).

The obvious fix is to change the line in `prepend` into

```rust
        let string_len = string.chars().count();
```

cc @estebank

### PR DESCRIPTION
**Pull request by AnthonyMikh**
Fix #65119 

r? @estebank

### CODE REVIEW

**Comment by AnthonyMikh** [comment]
...And let's just hope that it will _not_ introduce one more stable-to-nightly regression.

File: src/librustc_errors/styled_buffer.rs [hunk 1]
```diff
@@ -111,7 +111,7 @@ impl StyledBuffer {
 
     pub fn prepend(&mut self, line: usize, string: &str, style: Style) {
         self.ensure_lines(line);
-        let string_len = string.len();
+        let string_len = string.chars().count();
```
Comment (estebank) [inline src/librustc_errors/styled_buffer.rs]: I would love to actually have a test that exercises this, but I recall having issues on some platforms with files with unicode chars in their paths. Could you test locally how this behave with compound characters, zero width characters and double width characters? In other parts of the codebase we use `UnicodeWidthChar::width(c).unwrap_or(0)` to properly align underlines, but in this case you're dealing with these as text buffers so what you did should be ok.

**Comment by estebank** [comment]
r=me after testing locally against the three cases requested.

**Comment by AnthonyMikh** [comment]
With combining character:

```
$ build/x86_64-unknown-linux-gnu/stage2/bin/rustc ../﻿ã.rs --error-format=short
../﻿ã.rs:2:14: error: unterminated double quote string
error: aborting due to previous error
```

With zero width character (there is a zero-width non-joiner between `no` and `join`):

```
$ build/x86_64-unknown-linux-gnu/stage2/bin/rustc ../﻿no‌join.rs --error-format=short
../﻿no‌join.rs:2:14: error: unterminated double quote string
error: aborting due to previous error
```

With full-width characters:

```
$ build/x86_64-unknown-linux-gnu/stage2/bin/rustc '../﻿﻿ｆｕｌｌ ＿ｗｉｄｔｈ.rs' --error-format=short
../﻿﻿ｆｕｌｌ＿ｗｉｄｔｈ.rs:2:14: error: unterminated double quote string
error: aborting due to previous error
```

With example from [Manish's blog](https://manishearth.github.io/blog/2017/01/14/stop-ascribing-meaning-to-unicode-code-points/):

```
$ build/x86_64-unknown-linux-gnu/stage2/bin/rustc ../﻿﻿ᄀᄀᄀ각ᆨᆨ.rs --error-format=short
error: couldn't read ../﻿﻿ᄀᄀᄀ각ᆨᆨ.rs: No such file or directory (os error 2)
error: aborting due to previous error
```

Kudos to @kpp for actually carrying out the check.

**Pull request merged**